### PR TITLE
feat(modelgen): support clearing outdated models using a feature flag

### PR DESCRIPTION
**Pull request by phani-srikar**
_Issue #, if available:_. #50 

_Description of changes:_
These changes allow the developers to automatically remove outdated model files upon schema update. 
These will be released with a Feature Flag so that existing customers will not have any breaking changes.

_How are these changes tested:_
using unit tests and TS, Dart sample apps. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen/tests/commands/models.test.js [hunk 1]
```diff
@@ -71,42 +71,128 @@ describe('command - models', () => {
...
+            const outputDirectory = path.join(MOCK_PROJECT_ROOT, OUTPUT_PATHS[frontend]);
+            const preExistingFileName = 'preexisting.txt';
+            const preExistingFilePath = path.join(outputDirectory, preExistingFileName);
+            const mockedFiles = {};
+            mockedFiles[schemaFilePath] = {
+                'schema.graphql': ' type SimpleModel { id: ID! status: String } '
+            };
+            mockedFiles[outputDirectory] = {};
+            mockedFiles[outputDirectory][preExistingFileName] = 'A pre-existing file to be cleared';
+            mock(mockedFiles);
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/models.test.js:169]: Could you rebase this PR and change the import to `mockFs` as the previous review?

File: packages/amplify-codegen/tests/commands/models.test.js [hunk 2]
```diff
@@ -71,42 +71,128 @@ describe('command - models', () => {
...
+
+            // assert model generation succeeds
+            expect(graphqlCodegen.codegen).toBeCalled();
+
+            // assert that codegen generated in correct place
+            expect(fs.readdirSync(outputDirectory).length).toBeGreaterThan(1);
+
+            // assert that the pre-existing file is retained
+            expect(fs.existsSync(preExistingFilePath)).toBeTruthy;
+        });
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/models.test.js:195]: These two unit tests have some duplicate code at the beginning. You could use a new test suite and `beforeEach` to include the duplicate parts. And also there is no need to loop all frontend types for clearing tests.

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/tests/commands/models.test.js (hunk 2 above)
Comment (phani-srikar) [reply packages/amplify-codegen/tests/commands/models.test.js]: Moved these to a separate test suite and refactored common mock setup to a method. Retained the tests for all frontends, as they have different output paths.

File: packages/amplify-codegen/tests/commands/models.test.js [hunk 3]
```diff
@@ -71,43 +72,148 @@ describe('command - models', () => {
...
+      expect(fs.readdirSync(outputDirectory).length).toEqual(1);
+      expect(fs.existsSync(preExistingFilePath));
+
+      await generateModels(MOCK_CONTEXT);
+
+      // assert model generation succeeds
+      expect(graphqlCodegen.codegen).toBeCalled();
+
+      // assert that codegen generated in correct place
+      expect(fs.readdirSync(outputDirectory).length).toBeGreaterThan(1);
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/models.test.js:199]: The previous test set the value to 0 but this is 1 instead. Is this intended change?

File: packages/amplify-codegen/tests/commands/models.test.js [hunk 4]
```diff
@@ -71,43 +72,148 @@ describe('command - models', () => {
...
+                    service: 'AppSync', 
+                    providerPlugin: 'awscloudformation',
+                    resourceName: MOCK_PROJECT_NAME
+                  }
+              ]
+          }
+      );
+    MOCK_CONTEXT.amplify.executeProviderUtils.mockReturnValue([]);
+    MOCK_CONTEXT.amplify.pathManager.getBackendDirPath.mockReturnValue(MOCK_BACKEND_DIRECTORY);
+    graphqlCodegen.codegen.mockReturnValue(MOCK_GENERATED_CODE);
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/models.test.js:141]: The beforeEach in two test suites(line52-69 and line125-142) seems to have the same code/mock values. Can you move the duplicate mocking to the beginning?

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/tests/commands/models.test.js (hunk 4 above)
Comment (phani-srikar) [reply packages/amplify-codegen/tests/commands/models.test.js:115]: Refactored it to a method. Thanks

File: packages/amplify-codegen/tests/commands/models.test.js (hunk 3 above)
Comment (phani-srikar) [reply packages/amplify-codegen/tests/commands/models.test.js:173]: This one tests `Should not clear the output directory before generation if cleanGeneratedModelsDirectory is false` and hence the pre-existing file should not be deleted.

**Review by AaronZyLee — APPROVED** [review]
LGTM 🐐

**Pull request merged**